specification/dsdl/attributes.tex

```diff
@@ -0,0 +1,193 @@
+
+In the following text, the term \emph{referring statement} denotes a statement
+containing an expression which refers to the offset attribute.
+The term \emph{bit length set} is defined in section \ref{sec:dsdl_bit_length_set}.
+
+The value of the attribute is a function of the field attribute declarations preceding the referring statement
+and the category of the containing data schema definition.
+
+If the current data schema definition belongs to the tagged union category,
+the referring statement must be located after the last field attribute definition.
+A field attribute definition following the referring statement renders the current definition invalid.
+For tagged unions, the value of the offset attribute is defined as:
+$$
+    \texttt{\_offset\_}_\text{union} =
+    \bigcup\limits_{i=1}^{n}
+    \left\{ \lceil{}\log_2 n\rceil{} + b : b \in B_i \right\}
+$$
+where $n$ is the number of fields defined in the current union definition,
+and $B_i$ is the bit length set of the data type of the field number $i$.
+
+If the current data schema definition does not belong to the tagged union category,
+the referring statement may be located anywhere within the current data schema definition.
+The value of the offset attribute is defined as:
+$$
+    \texttt{\_offset\_} =
+    \begin{cases}
+        \left\{ \sum s : s \in \prod\limits_{i=1}^{n} B_i \right\} &\mid n > 0 \\
+        \left\{0\right\}                                           &\mid n = 0 \\
+    \end{cases}
+$$
+where $n$ is the number of fields defined in statements preceding the referring statement
+(see section \ref{sec:dsdl_grammar} on statement ordering),
+$B_i$ is the bit length set of the data type of the field number $i$,
+and $\prod$ is the Cartesian product operator.
+
+\begin{remark}
+    \begin{minted}{python}
+        @union
+        uint8 a
+        #@print _offset_  # Would fail: it's a tagged union, _offset_ is undefined until after the last field
+        uint16 b
+        @assert _offset_ == {1 + 8,  1 + 16}
+        ---
+        @assert _offset_ == {0}
+        float16 a
+        @assert _offset_ == {16}
+        void4
+        @assert _offset_ == {20}
+        int4 b
+        @assert _offset_ == {24}
+        uint8[<4] c
+        @assert _offset_ == 2 + {24,  32,  40,  48}
+        void6
+        # One of the main usages for _offset_ is statically proving that the following field is byte-aligned
+        # for all possible valid serialized representations of the preceding fields. It is done by computing
+        # a remainder as shown below. If the field is aligned, the remainder set will equal {0}. If the
+        # remainder set contains other elements, the field may be misaligned under some circumstances.
+        # If the remainder set does not contain zero, the field is never aligned.
+        @assert _offset_ % 8 == {0}
+        uint8 well_aligned   # Proven to be byte-aligned.
+    \end{minted}
+\end{remark}
```

Comment:
__offset__ is sort of complex. Won't most authors really care about the maximum size anyway? Should we provide __max_offset__ that is a simple scalar value?

Reply:
As discussed, `_offset_` is crucial for static alignment analysis (and https://github.com/UAVCAN/pydsdl/pull/24, although it has no direct relation to the specification itself). The case of `_max_offset_` is also supported as `_offset_.max`.